receiver/otlpreceiver/otlp.go

```diff
@@ -230,3 +230,7 @@ func (r *otlpReceiver) registerLogsConsumer(ctx context.Context, tc consumer.Log
 	}
 	return nil
 }
+
+func (r *otlpReceiver) SetServiceExtensions(extensions map[string]component.ServiceExtension) {
+	r.cfg.GRPC.SetServiceExtensions(extensions)
```

Comment:
Can't we auto-inject directly into the configgrpc? The goal is to remove this knowledge from individual components.

Reply:
The issue I ran into with trying to inject into the configgrpc, is that there's no guaranteed place for it on an individual component's config. For example, the [otlp receiver has the GRPC server settings object inside a `Protocols` struct](https://github.com/open-telemetry/opentelemetry-collector/blob/a16c599d26b6db7a95489568ca890b06380463cc/receiver/otlpreceiver/config.go#L23) on the config, while the [opencensus receiver just has the GRPC settings as top level attributes on its config struct](https://github.com/open-telemetry/opentelemetry-collector/blob/master/receiver/opencensusreceiver/config.go#L27). This led me to writing a recursive reflection function to iterate over the entire config struct which to me feels a bit messy and overcomplicated.

Reply:
Thinking one abstraction further, perhaps the unmarshaller could check whether the struct it just created implements the new interface?

Reply:
The config yaml is unmarshalled before any extensions are built, which makes sense because the extensions need their own config to be built

Reply:
You can keep a record of the objects that were built and desire extensions. You can iterate this list later and inject the desired requests.

Reply:
I don't think getting this information at the time the config is unmarshalled is as straightforward as it seems. After looking through the service code a bit, it seems the service seems can be configured to use a different ConfigFactory at runtime, which any config factory would need to know to return this list of configs that want service extensions. 

Additionally it seems like individual components can be configured with a custom unmarshall function, which it doesn't seem we want want to impose required logic into. 

That seems to leave us back in the place where we would need to reflect over the fully loaded config struct to determine if any portion of it wants the service extensions.

Forgive me if I'm missing an obvious place in the codebase that can be leveraged here.

Reply:
You are probably not missing, I haven't looked at this code for a while. I'll take a look in early Jan to see if I have any new ideas.

Reply:
@jpkrohling I think the [draft you posted](https://github.com/open-telemetry/opentelemetry-collector/compare/master...jpkrohling:jpkrohling/2229-inject-authenticator#diff-85ee04e67bda04e6159b92426c87fe161c58d4b5550ef2b0fa7a97214309c125R378) currently leads to this same issue, where the built exporter would still need to be responsible for passing the authenticator to the grpc config settings, instead of being able to push it straight into the grpc config.

Reply:
True, but the config model is the key there. I think it's possible to inject the authenticator to the config. I'll take a look tomorrow.

Reply:
The code editor was still open and the idea fresh in my head, so, I went ahead and pushed another commit, injecting the authenticator in the exporter's config. But you are right, something (exporter? its config?) would still need to make the bridge between the gRPC settings and the authenticator. I never thought I would say this, but I miss Java right now, especially CDI :-)

Reply:
In time: we just missed the train, but I think this would be a good topic for the SIG meeting, to see if other folks have other ideas. https://docs.google.com/document/d/1r2JC5MB7GupCE7N32EwGEXs9V_YIsPgoFiLP4VWVMkE